### PR TITLE
Implement multiple client certificate validation subjects

### PR DESCRIPTION
**Pull request by radekg**
Hey @everesio. This is the PR which changes the client certificate validation implementation. It allows specifying multiple subjects. When at least one certificate passes at least one subject under test, the request will be allowed.

The possible values are:

- simple string matching: `s:/FIELD=[value,value,...]/FIELD2=[....]/...`
- pattern matching: `r:/FIELD=[^regex.*$,^regex2.{1,3}$,...]/FIELD2=[....]/...`

Hand rolled parser makes sure that:

- prefix is correct (`s:` or `r:`)
- fields are separated by `/`
- field names are uppercase alpha strings followed by `=`
- values are lists enclosed within `[...]` and always having closing `]`
- values are comma separated, in case of regex values, the parser makes sure the split happens only within the top `[...]`, i.e. nested `[]` are allowed and commas inside of `{}` do not split values
- nested `[]` are not allowed for string type matching

Comprehensive tests with `makeTLSPipe` are in.

**Requires readme updates:** if this looks okay to you, I'll make updates to the readme and it will be ready to merge.

### CODE REVIEW

**Comment by radekg** [comment]
I don't like the `ParsedSubject.Validate() error` method. I'm going to incorporate it in the parser.

**Review by everesio — CHANGES_REQUESTED** [review]
Hi @radekg  
thank you for your PR. I've just added some comments
Kind regards,
Michal

File: proxy/tls_client_cert_validate.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		if len(parsedSubjects) == 0 {
+			return nil // nothing to validate
+		}
+
+		errs := []error{}
+
+		for _, chain := range verifiedChains {
+			for _, cert := range chain {
+
+				if len(parsedSubjects) == 0 {
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/tls_client_cert_validate.go]: Resolved in: https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e#diff-ecd926b7513cf5217a1c45eadfb7ba725252f6f448f3ef962b584eef927f0eceL27.

File: proxy/clientcertvalidate/parser.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		n = p.maxPos - p.pos
+	}
+	if n == 0 {
+		return []rune{}, io.EOF
+	}
+	defer func() { p.pos = p.pos + n }()
+	return []rune(p.input[p.pos : p.pos+n]), err
+}
+
+func (p *defaultSubjectParser) skip(n int) {
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/parser.go:56]: Resolved in https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e#diff-ecd926b7513cf5217a1c45eadfb7ba725252f6f448f3ef962b584eef927f0eceL27.

File: proxy/clientcertvalidate/parser.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+			output.kvs[field] = values
+		}
+
+		return output, nil
+	}
+
+	return output, &ParserInvalidPrefixError{ConsumedPrefix: string(prefix)}
+}
+
+func (p *defaultSubjectParser) readAlphaStringUntil(boundary rune) (string, error) {
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/parser.go:198]: Resolved in https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e#diff-ecd926b7513cf5217a1c45eadfb7ba725252f6f448f3ef962b584eef927f0eceL27.

File: proxy/clientcertvalidate/parser.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+	consumed := []rune{}
+	for {
+		nextRune, err := p.lookupOne()
+
+		if err == io.EOF {
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/parser.go]: Added better error handling. https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e#diff-ecd926b7513cf5217a1c45eadfb7ba725252f6f448f3ef962b584eef927f0eceL27

File: proxy/clientcertvalidate/parser.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+			return values, io.EOF
+		}
+
+		if nextRune == '[' {
+			nest = nest + 1
+			if nest > 1 {
+				if inputValuesType == ClientCertificateSubjectPrefixString {
+					// a string type subject can't contain nested patterns:
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/parser.go]: I've reworked the parser to allow these. Indeed, makes it less complex. https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e#diff-ecd926b7513cf5217a1c45eadfb7ba725252f6f448f3ef962b584eef927f0eceL27

File: proxy/clientcertvalidate/parser.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+			}
+			nest = nest - 1
+		}
+
+		// values are separated by commas but they can only exist when we are inside
+		// top level [...] and not between {...}
+		if nextRune == '{' {
+			// nested {} are not okay...
+			if insidePattern {
+				return values, errNestedPattern
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/parser.go]: Parser is now much simpler with this update: https://github.com/grepplabs/kafka-proxy/pull/74/commits/82d4f4b85d2cdd34b6a418b01bd83c5a475fdd4e.

File: proxy/tls_client_cert_validate.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+					return nil
+				}
+
+				certificateAcceptable := false
+
+				// as soon as any parsed subject validates, pass the request;
+				for _, parsedSubject := range parsedSubjects {
+					x509ValidateErr := parsedSubject.X509Validate(cert)
+					if x509ValidateErr == nil {
+						certificateAcceptable = true
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/tls_client_cert_validate.go]: Resolved with https://github.com/grepplabs/kafka-proxy/pull/74/commits/8d0c7df87181cae5177bcf7ea7ead5ba34f80fb8.

File: proxy/clientcertvalidate/subject.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+			}
+		}
+		return nil
+	}
+	if headerType == ClientCertificateSubjectPrefixPattern {
+		if len(expected) != len(certValues) {
+			return fmt.Errorf("%s: expected has %d elements vs %d in certificate", field, len(expected), len(certValues))
+		}
+		for index, certValue := range certValues {
+			matched, err := regexp.Match(expected[index], []byte(certValue))
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/subject.go:124]: Indeed, an I am already compiling the patter after parse. I should hold on to the compiled one and reuse it. Let me rework this.

[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/subject.go]: Resolved with https://github.com/grepplabs/kafka-proxy/pull/74/commits/50651965ea825a1d3856e4c8517e008cfbb125f8/

File: proxy/clientcertvalidate/subject.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+func testCertValuesSlices(field string, expected, certValues []string, headerType ClientCertificateSubjectPrefixType) error {
+	if headerType == ClientCertificateSubjectPrefixString { // easy: just sort and check if true:
+		sort.Strings(expected)
```
[the remark this answers is not in the document]
Comment (radekg) [reply proxy/clientcertvalidate/subject.go]: Resolved in https://github.com/grepplabs/kafka-proxy/pull/74/commits/9f5ba378bf36e5337a5e50bdb01f407ad7579d47.

**Comment by radekg** [comment]
Hey @everesio. Thanks for the review. I've updated the PR to include your comments. One note regarding `readEscapeSequence`. This function does not try reading hex strings, octal number, nor unicode sequences. It could do so but at the moment it does not.

File: proxy/clientcertvalidate/subject.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+			}
+		}
+	default:
+		return errors.New("unknown subject header")
+	}
+
+	return nil
+}
+
+func testCertValuesSlicesStringSubject(field string, expected, certValues []string) error {
```
Comment (everesio) [inline proxy/clientcertvalidate/subject.go:163]: field is not used, clientCertSubjectOrganizationalUnit is hard coded

Comment (radekg) [reply proxy/clientcertvalidate/subject.go:163]: Thanks, fixed: https://github.com/grepplabs/kafka-proxy/pull/74/commits/0ff634d1ff1902a75b36704db539b2c76cdd95d7.

**Pull request merged**